### PR TITLE
feat(Core/DB): port WarheadCore database

### LINKED ISSUE #14572: [server]world server crash 0109
### Current Behaviour

crash log:
https://gist.github.com/mpfans/868911f0fbc204760d4ccff642c7f4d9

### Expected Blizzlike Behaviour

none

### Source

_No response_

### Steps to reproduce the problem

none

### Extra Notes

_No response_

### AC rev. hash/commit

8d2dda24ed85e835953156d1df429567bfc6fea8

### Operating system

Ubuntu20.04

### Custom changes or Modules

_No response_

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## Changes Proposed:
- Added support dynamic connections
- Using base class for connections instead of template
- Using smart poiner for `PreparedStatement`

## Issues Addressed:
- Maybe close #14572

## SOURCE:
- https://github.com/WarheadCore/WarheadBand

## Tests Performed:
- Build, in-game
- This changes tested in live servers with 10-60 playes (With WH core)

## How to Test the Changes:
- Need a stress test of these changes on the server with the players

## Deps modules
- https://github.com/azerothcore/mod-progression-system/pull/340
- https://github.com/azerothcore/mod-eluna/pull/148
- https://github.com/azerothcore/mod-chat-transmitter/pull/7
- https://github.com/azerothcore/mod-anticheat/pull/122 (Can merge before this pr)
- https://github.com/azerothcore/mod-transmog/pull/127 (Merged)

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/database/Database/DatabaseAsyncQueueWorker.h [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef WARHEAD_ASYNC_DB_QUEUE_WORKER_H_
+#define WARHEAD_ASYNC_DB_QUEUE_WORKER_H_
```
Comment (Kitzunu) [inline src/server/database/Database/DatabaseAsyncQueueWorker.h:19]: ^

**Comment by Kitzunu** [comment]
What is the benefit for this?

**Comment by Winfidonarleyan** [comment]
> What is the benefit for this?

Perfomace

File: src/server/database/Database/DatabaseAsyncQueueWorker.h (hunk 1 above)
Comment (Winfidonarleyan) [reply src/server/database/Database/DatabaseAsyncQueueWorker.h:19]: ah yes, ty

**Comment by Kitzunu** [comment]
We love performance 😎

**Comment by Winfidonarleyan** [comment]
> We love performance sunglasses

Everyone loves it, but it takes a lot of testing.

**Comment by Kitzunu** [comment]
> And now can make new db in modules (need test anyway) :D

Now this is cool

**Comment by Kitzunu** [comment]
That should be documented though imo

**Comment by Winfidonarleyan** [comment]
> That should be documented though imo

Absolutely, have to test everything first.

**Comment by heyitsbench** [comment]
Modules gonna need updating. 😬

**Comment by Winfidonarleyan** [comment]
> Modules gonna need updating. 😬

yes yes

**Comment by Winfidonarleyan** [comment]
Update modules

File: src/server/database/Database/DatabaseWorkerPool.cpp [hunk 2]
```diff
@@ -16,248 +16,428 @@
...
+    bool isSupportClientDB = mysql_get_client_version() >= MIN_DB_CLIENT_VERSION;
     bool isSameClientDB    = true; // Client version 3.2.3?
 #endif
 
-    WPFatal(isSupportClientDB, "AzerothCore does not support MySQL versions below 5.7 or MariaDB versions below 10.5.\n\nFound version: {} / {}. Server compiled with: {}.\nSearch the wiki for ACE00043 in Common Errors (https://www.azerothcore.org/wiki/common-errors#ace00043).",
-        mysql_get_client_info(), mysql_get_client_version(), MYSQL_VERSION_ID);
-    WPFatal(isSameClientDB, "Used MySQL library version ({} id {}) does not match the version id used to compile AzerothCore (id {}).\nSearch the wiki for ACE00046 in Common Errors (https://www.azerothcore.org/wiki/common-errors#ace00046).",
-        mysql_get_client_info(), mysql_get_client_version(), MYSQL_VERSION_ID);
+    ASSERT(isSupportClientDB, "AzerothCore does not support MySQL versions below 5.7 and MariaDB 10.3");
+    ASSERT(isSameClientDB, "Used DB library version ({} id {}) does not match the version id used to compile AzerothCore (id {})", mysql_get_client_info(), mysql_get_client_version(), MYSQL_VERSION_ID);
```
Comment (Kitzunu) [inline src/server/database/Database/DatabaseWorkerPool.cpp:84]: \nSearch the wiki for ACE00043 in Common Errors (https://www.azerothcore.org/wiki/common-errors#ace00043)
\nSearch the wiki for ACE00046 in Common Errors (https://www.azerothcore.org/wiki/common-errors#ace00046).

Comment (Winfidonarleyan) [reply src/server/database/Database/DatabaseWorkerPool.cpp:84]: Ah yes, thanks

Comment (Winfidonarleyan) [reply src/server/database/Database/DatabaseWorkerPool.cpp]: Added

**Comment by walkline** [comment]
I understand that this might exceed the current scope, and it's more like my stream of consciousness. But if we could connect to multiple character databases, bind them to players, and incorporate PR with clustering support, we might be close to a cross-realm implementation.

**Comment by Kitzunu** [comment]
Status?

**Comment by Winfidonarleyan** [comment]
Need time :D

**Comment by Nyeriah** [comment]
Please reopen when continue

**Pull request closed without merging**